### PR TITLE
feat(#259): add OAuth application logic (Epic 3)

### PR DESCRIPTION
**Pull request by Derane**
## Summary

- **ST-3.1**: `OAuthUserResolver` with 3 resolution paths — existing social identity (touch lastUsed), auto-link by email (confirm unconfirmed users), and new user creation (random hashed password, confirmed=true)
- **ST-3.2/3.3**: `InitiateOAuthCommand`/`HandleOAuthCallbackCommand` with handlers implementing full OAuth flow — state+PKCE generation, atomic state validation, code exchange, user resolution, 2FA branching via extracted `OAuthCallbackTwoFactorHandler`
- **ST-3.4**: Domain events (`OAuthUserCreatedEvent`, `OAuthUserSignedInEvent`) extending `OAuthDomainEvent`, event factory, and publisher wrapping EventBus
- **ST-3.5**: Service wiring in `services.yaml` for resolver, event factory, publisher, and handler TTL config
- 100% unit test coverage across all 16 new production classes (16 test files, 2068 total tests passing)

## Architecture decisions

- Extracted `OAuthCallbackTwoFactorHandler` to keep `HandleOAuthCallbackCommandHandler` under PHPMD CBO threshold (13)
- Used `EventIdFactoryInterface` instead of `UuidFactory` in `OAuthUserResolver` to reduce coupling
- Used `new User()` directly instead of `UserFactoryInterface` in resolver to stay under CBO limit
- All new classes annotated with `@psalm-api` since consumers (Epic 4 controllers) don't exist yet

## Test plan

- [x] All 2068 unit tests pass with 100% line coverage
- [x] Psalm: 0 errors
- [x] Deptrac: 0 violations  
- [x] PHPInsights: architecture, style, complexity, quality all pass
- [x] PHPMD: 0 violations
- [x] PHP CS Fixer: clean
- [ ] `validate-configuration` fails on `psalm.xml` — inherited from parent Epic 2b branch, not from this PR's changes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added full OAuth sign-in flow (initiate + callback), automatic account linking/creation, two-factor branch with pending-session support, and event publishing for user creation/sign-in.

* **Configuration**
  * Added Facebook Graph API version setting for OAuth providers.

* **Tests**
  * Comprehensive unit test coverage for OAuth flows, handlers, resolvers, DTOs, events, factories, and publishers.

### CODE REVIEW

**Comment by Derane** [comment]
@coderabbitai full review

File: src/OAuth/Application/Command/InitiateOAuthCommand.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function getResponse(): InitiateOAuthResponse
+    {
+        return $this->response;
+    }
+
+    public function setResponse(InitiateOAuthResponse $response): void
+    {
+        $this->response = $response;
+    }
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/Command/InitiateOAuthCommand.php:31]: This follows the existing command pattern used across all commands in the project (e.g., `CompleteTwoFactorCommand`, `RefreshTokenCommand`, `SignInCommand`, etc.) — all use non-nullable `$response` properties with getter/setter. The response-on-command pattern is the established convention here, and changing only this command would create inconsistency. Moving to command-in/response-out would be a broader project-level refactor.

File: src/OAuth/Application/CommandHandler/OAuthCallbackTwoFactorHandler.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        HandleOAuthCallbackCommand $command,
+    ): void {
+        $now = new DateTimeImmutable();
+        $pending = $this->pendingTwoFactorFactory->create(
+            $this->idFactory->create(),
+            $user->getId(),
+            $now,
+            $now->modify(
+                sprintf('+%d seconds', $this->pendingTwoFactorTtlSeconds)
+            ),
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/CommandHandler/OAuthCallbackTwoFactorHandler.php:49]: Good catch. Added constructor validation that throws `InvalidArgumentException` when `pendingTwoFactorTtlSeconds <= 0`. Also added unit tests for zero and negative TTL values.

File: src/OAuth/Application/DTO/HandleOAuthCallbackResponse.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public function getRefreshToken(): ?string
+    {
+        return $this->refreshToken;
+    }
+
+    public function getPendingSessionId(): ?string
+    {
+        return $this->pendingSessionId;
+    }
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/DTO/HandleOAuthCallbackResponse.php:38]: This mirrors `SignInCommandResponse` exactly (same private properties with getters pattern). The guideline about public properties applies to request DTOs validated via `config/validator/validation.yaml` — this is a response DTO, not a request DTO. All existing response DTOs in the project use private properties with getters.

File: src/OAuth/Application/Resolver/OAuthUserResolver.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+use App\User\Application\Factory\IdFactoryInterface;
+use App\User\Domain\Contract\PasswordHasherInterface;
+use App\User\Domain\Entity\User;
+use App\User\Domain\Repository\UserRepositoryInterface;
+use DateTimeImmutable;
+
+/**
+ * @psalm-api
+ */
+final readonly class OAuthUserResolver implements OAuthUserResolverInterface
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/Resolver/OAuthUserResolver.php:22]: `Application/Resolver/` is not GraphQL-only in this project. It already contains non-GraphQL resolvers: `BatchEmailResolver`, `HttpExceptionDetailResolver`, `QueryViolationResolver`, etc. The directory name matches the well-known Resolver pattern — it resolves which user corresponds to an OAuth profile. No move needed.

File: src/OAuth/Application/Resolver/OAuthUserResolver.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            return $this->handleExistingIdentity($existing);
+        }
+
+        $user = $this->userRepository->findByEmail($profile->email);
+
+        if ($user instanceof User) {
+            return $this->handleAutoLink($user, $provider, $profile);
+        }
+
+        return $this->handleNewUser($provider, $profile);
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/Resolver/OAuthUserResolver.php:54]: Valid security concern. Fixed: `resolve()` now skips the auto-link-by-email path when `$profile->emailVerified` is false (goes straight to new user creation), and `createOAuthUser()` only calls `setConfirmed(true)` when email is verified. Added two unit tests covering both behaviors.

File: src/OAuth/Application/Resolver/OAuthUserResolver.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        $this->createAndSaveIdentity(
+            $provider,
+            $profile->providerId,
+            $user->getId(),
+        );
+
+        if (!$user->isConfirmed()) {
+            $user->setConfirmed(true);
+            $this->userRepository->save($user);
+        }
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Application/Resolver/OAuthUserResolver.php:86]: MongoDB does not support multi-collection transactions without replica sets, and this project uses standalone MongoDB for development/testing. The existing codebase follows the same pattern — separate `save()` calls across repositories (e.g., in `SignInCommandHandler`, `ConfirmTwoFactorCommandHandler`). The emailVerified gate added above eliminates the most critical risk (linking to unverified emails). Race conditions on the identity lookup are mitigated by the unique index on `(provider, providerId)`.

File: src/OAuth/Domain/Event/OAuthUserSignedInEvent.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+     *
+     * @return self
+     */
+    #[\Override]
+    public static function fromPrimitives(
+        array $body,
+        string $eventId,
+        string $occurredOn
+    ): DomainEvent {
+        return new self(
```
[the remark this answers is not in the document]
Comment (Derane) [reply src/OAuth/Domain/Event/OAuthUserSignedInEvent.php:33]: All existing domain events in this project use `: DomainEvent` as the return type for `fromPrimitives()` — e.g., `UserSignedInEvent`, `UserRegisteredEvent`, `PasswordChangedEvent`, etc. This is consistent with the established convention. Changing only these events would create inconsistency.

File: tests/Unit/OAuth/Application/CommandHandler/HandleOAuthCallbackCommandHandlerTest.php [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+            $this->providerName,
+            $codeVerifier ?? $this->faker->sha256(),
+            $this->faker->sha256(),
+            $this->faker->url(),
+            new DateTimeImmutable(),
+        );
+
+        $this->stateRepository->method('validateAndConsume')
+            ->willReturn($payload);
+    }
```
[the remark this answers is not in the document]
Comment (Derane) [reply tests/Unit/OAuth/Application/CommandHandler/HandleOAuthCallbackCommandHandlerTest.php:204]: Good point. Updated `arrangeStatePayload()` to assert the provider name and type-check the state/flowBindingToken arguments via `->with()` constraints.

File: tests/Unit/OAuth/Infrastructure/Publisher/OAuthPublisherTest.php [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        $email = $this->faker->safeEmail();
+        $provider = $this->faker->word();
+        $sessionId = $this->faker->uuid();
+        $eventId = $this->faker->uuid();
+
+        $event = $this->arrangeSignedInEvent($eventId);
+
+        $this->oAuthEventFactory->expects($this->once())
+            ->method('createUserSignedIn')
+            ->willReturn($event);
```
[the remark this answers is not in the document]
Comment (Derane) [reply tests/Unit/OAuth/Infrastructure/Publisher/OAuthPublisherTest.php]: Fixed. The test now constructs the event with the same variables passed to `publishUserSignedIn()` and asserts the factory receives the correct `$userId`, `$email`, `$provider`, `$sessionId`, and `$eventId` via `->with()`.

**Review by dmytrocraft — DISMISSED** [review]
Merged the latest main into this branch, resolved the resulting conflicts, and verified the merge-critical path locally with the affected OAuth unit tests plus a Symfony container lint.

**Review by dmytrocraft — DISMISSED** [review]
Addressed the follow-up review points on the current head: added a guard for dangling OAuth identity links, tightened publisher test interaction expectations, and added failure-path coverage plus focused mutation verification.

**Review by dmytrocraft — DISMISSED** [review]
Re-approving after CI fixups.

**Review by dmytrocraft — DISMISSED** [review]
Re-approving after PHP Insights follow-up.

**Review by dmytrocraft — APPROVED** [review]
Re-approving after final PHP Insights cleanup and main merge resolution.

**Review by dmytrocraft — APPROVED** [review]
Re-approving after dismissing the non-blocking bot review; all required checks are green.

**Review by Kravalg — APPROVED** [review]
